source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,412 @@
+       "text/x-js", "application/ecmascript", "application/x-json", "application/xml",
+       "application/json", "image/svg+xml", "text/xml", "application/xhtml+xml"});
+}
+
+// List of CompressorFilterConfig objects registered for a stream.
+struct CompressorRegistry : public StreamInfo::FilterState::Object {
+  std::list<CompressorFilterConfigSharedPtr> filter_configs_;
+};
+
+// Key to per stream CompressorRegistry objects.
+const std::string& compressorRegistryKey() { CONSTRUCT_ON_FIRST_USE(std::string, "compressors"); }
+
+} // namespace
+
+CompressorFilterConfig::CompressorFilterConfig(
+    const envoy::extensions::filters::http::compressor::v3::Compressor& compressor,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+    const std::string& content_encoding)
+    : content_length_(contentLengthUint(compressor.content_length().value())),
+      content_type_values_(contentTypeSet(compressor.content_type())),
+      disable_on_etag_header_(compressor.disable_on_etag_header()),
+      remove_accept_encoding_header_(compressor.remove_accept_encoding_header()),
+      stats_(generateStats(stats_prefix, scope)), runtime_(runtime),
+      content_encoding_(content_encoding) {}
+
+StringUtil::CaseUnorderedSet
+CompressorFilterConfig::contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types) {
+  const auto& default_content_encodings = defaultContentEncoding();
+  return types.empty() ? StringUtil::CaseUnorderedSet(default_content_encodings.begin(),
+                                                      default_content_encodings.end())
+                       : StringUtil::CaseUnorderedSet(types.cbegin(), types.cend());
+}
+
+uint32_t CompressorFilterConfig::contentLengthUint(Protobuf::uint32 length) {
+  return length > 0 ? length : DefaultMinimumContentLength;
+}
+
+CompressorFilter::CompressorFilter(const CompressorFilterConfigSharedPtr config)
+    : skip_compression_{true}, config_(std::move(config)) {}
+
+Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  if (accept_encoding != nullptr) {
+    accept_encoding_ = std::make_unique<std::string>(accept_encoding->value().getStringView());
+  }
+
+  if (config_->runtime().snapshot().featureEnabled(config_->featureName(), 100)) {
+    skip_compression_ = false;
+    if (config_->removeAcceptEncodingHeader()) {
+      headers.removeAcceptEncoding();
+    }
+  } else {
+    config_->stats().not_compressed_.inc();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void CompressorFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+
+  absl::string_view key = compressorRegistryKey();
+  const StreamInfo::FilterStateSharedPtr& filter_state = callbacks.streamInfo().filterState();
```

Comment:
Please add more comment throughout this file. Here, for example, please comment on what the filter state is used for.